### PR TITLE
Remove deprecated objective functions and target layer

### LINKED ISSUE #193: A proposal for a new objective function paradigm
As we move beyond classical problems like ImageNet classification, we are starting to develop a need for more problem-specific objective functions. We currently have to implement a new derived class for every objective function term, so experimenting with objective functions is slow, bug-prone, and contributes to feature bloat. It can also lead to confusing situations. For instance, our usage of KL divergence in variational autoencoders is non-standard and the implementation is completely different from that Keras and Tensorflow. [Explanation: Typically KL divergence is calculated between two distributions, e.g. probability vectors for "ground truth" and "prediction." However, VAEs compute the KL divergence caused by adding noise. The inputs are not probability vectors, but rather the noise distribution parameters.]

I think there is better way. Instead of building and maintaining these complicated objective_function_term objects, we should instead support a collection of layers to perform simple entry-wise operations and column-wise reductions. For instance, the KL divergence term in VAEs can be computed with exponentiation and addition (both of which we already support), followed by a column-wise average. This will result in a 1xN activations matrix where each entry is the KL divergence corresponding to a mini-batch sample. This can be output to an objective_function_layer that will be responsible for extracting the values and initializing gradients for backprop.

This proposal would eliminate objective_function_term classes corresponding to loss functions and activation regularization. However, it would be a good idea to implement layers for common loss functions like cross entropy or mean squared error.

### LINKED ISSUE #678: Cull deprecated features
### Deprecated features
- [x] Target layers
- [x] Reconstruction layers
- [x] All metrics except layer metric
- [x] All objective function terms except layer term and L2 regularization term
- [x] All models except DAG model

### Motivation
The target-layer-based objective function paradigm has been deprecated for months by the evaluation layer paradigm (#193). The target layer paradigm is somewhat simpler for models that fit into a rigid classification/regression/autoencoder framework, but is impractical if the objective function has any complexity. The evaluation layer paradigm is also more convenient from a code maintenance standpoint and can take advantage of GPU support in the layer class. We've supported both paradigms for backward compatibility, but this also makes it difficult to make substantial changes to our core objective function machinery. This difficulty is blocking progress on some performance optimizations (#632). I would say that this is a good point to bite the bullet and just remove our support for the deprecated features. Since most of our specialized model classes rely on the target layer paradigm, we should remove them as well.

### Progress
I've made a series of PRs to remove deprecated functionality from the model zoo (#667, #668, #669, #671, #675, #676). Once these PRs are merged, I believe the only offending models will be the Siamese models (pinging @JaeseungYeom). If anyone has models that aren't in the model zoo, they should check for deprecated functionality before we officially stop supporting them. For an example, I would recommend looking at how we updated LeNet (2376c0027bf75c36db553006edc0b1aec35f3fd2).

### Recommendations
- Change model type to `directed_acyclic_graph_model` (or remove the `name` field since DAG models are now default). If the model was previously a `sequential_model`, this will require specifying the parent/child relationships between layers.
- Specify two identity layers (or split layers or dummy layers) as children of the input layer. Input layers expect two child layers (one for sample data and one for "ground truth" data), so being explicit helps make sure you're getting the data you expect from the input layer. An example:
```
layer {
  name: "input"
  children: "image label"
  input {
    io_buffer: "partitioned"
  }
}
layer {
  parents: "input"
  name: "image"
  identity {}
}
layer {
  parents: "input"
  name: "label"
  identity {}
}
```
- Remove target and reconstruction layers and replace them with the appropriate loss layers. In most cases, they will require two parents like the target layer.
- Replace any terms in the objective function (except L2 weight regularization) with:
```
layer_term { layer: "name_of_corresponding_loss_layer" }
```
- Replace any metrics with:
```
metric {
  metric_term {
    name: "human-readable metric name"
    layer: "name_of_corresponding_loss_layer"
#   unit: "%" # Include if you want to display a percentage, e.g. for accuracy
  }
}
```

### PR DESCRIPTION
**Pull request by timmoon10**
This eliminates the last vestiges of the old target-layer-based objective function paradigm. This should be fairly painless since we already switched to the layer-based objective function paradigm several months ago (see Issue #678). All the gradient checking unit tests run and pass. AlexNet also runs (1 Pascal node, 1% of ILSVRC-2012) with no substantial change to learning behavior or performance.

This closes #193 and closes #678.

### CODE REVIEW

**Comment by timmoon10** [comment]
Replacements for deprecated objective functions, using the Python interface in https://github.com/ndryden/lbann/pull/1:
- `binary_cross_entropy`: `BinaryCrossEntropy([pred, label])`
- `cross_entropy`: `CrossEntropy([pred, label])`
- `cross_entropy_with_uncertainty`:
```python
def cross_entropy_with_uncertainty(pred, label):
    masked_pred = Multiply([pred, label]) # Assuming `label` is Boolean
    pred_sum = Reduction(masked_pred)
    return Negative(Log(pred_sum))
```
- `geom_negloglike`:
```python
def geom_negloglike(pred, label):
    ones = Constant(hint_layer=pred, value=1.0)
    term1 = Multiply([label, Log(Subtract([ones, pred]))])
    term2 = Log(pred)
    full = WeightedSum([term1, term2], scaling_factors='-1.0 -1.0')
    return Reduction(full)
```
- `mean_absolute_deviation`: `MeanAbsoluteError([pred, label])`
- `mean_absolute_error`: `MeanAbsoluteError([pred, label])`
- `mean_squared_error`: `MeanSquaredError([pred, label])`
- `poisson_negloglike`:
```python
def poisson_negloglike(pred, label): # Requires log-gamma, which is not yet supported
    ones = Constant(hint_layer=pred, value=1.0)
    term1 = pred
    term2 = Multiply([label, Log(pred)])
    term3 = LogGamma(Add([label, ones]))
    full = WeightedSum([term1, term2, term3], scaling_factors='1.0 -1.0 1.0')
    return Reduction(full)
```
- `polya_negloglike`:
```python
def polya_negloglike(pred, label): # Requires log-gamma, which is not yet supported
    count = Reduction(label)
    alpha_sum = Reduction(pred)
    lgamma_alpha_sum = Reduction(LogGamma(pred)) 
    lgamma_alpha_level_count_sum = Reduction(LogGamma(Add([pred, label])))
    return WeightedSum([LogGamma(alpha_sum),
                        LogGamma(Sum([count, alpha_sum])),
                        lgamma_alpha_level_count,
                        lgamma_alpha_sum],
                       scaling_factors='-1.0 1.0 -1.0 1.0'])

```
- `group_lasso_weight_regularization`:
```python
def group_lasso_weight_regularization(weights, height, width):
    w = WeightsLayer(weights=weights, dims='%d %d'.format(width, height))
    slice = Slice(w, slice_axis=0, slice_points=' '.join(range(width+1)))
    cols = []
    for _ in range(width):
        cols.append(Sqrt(L2Norm2(slice)))
    return Sum(cols)
```
- `l1_weight_regularization`:
```python
def l1_weight_regularization(weights, dims):
    w = WeightsLayer(weights=weights, dims=' '.join(dims))
    return L1Norm(w)
```

**Comment by bvanessen** [comment]
@timmoon10 Can you please put these functions into the new python module library.

**Comment by timmoon10** [comment]
I'm a bit wary about including the negative log-likelihoods. As far as I can tell, they're not natively supported in PyTorch, Keras, or Caffe (you might be able to compute log-likelihoods for geometric and Poisson distributions in TensorFlow if you delve into their support for statistical distributions, but I'm not sure). These are expert functionalities for very statistically sophisticated models and I don't think it will be worthwhile to maintain such specialized operations.

I think group LASSO has a similar problem in that it's still a rather exotic technique that isn't supported by anyone else. The L1 regularization seems common enough, though, to warrant including in the Python interface.
